src/data/padics/padic_integers.lean

```diff
@@ -246,30 +291,61 @@ show ∥((p : ℤ_[p]) : ℚ_[p])∥ = p⁻¹, by exact_mod_cast padic_norm_e.no
 show ∥((p^n : ℤ_[p]) : ℚ_[p])∥ = p^(-n:ℤ),
 by { convert padic_norm_e.norm_p_pow n, simp, }
 
-end padic_norm_z
+private def cau_seq_to_rat_cau_seq (f : cau_seq ℤ_[p] norm) :
+  cau_seq ℚ_[p] (λ a, ∥a∥) :=
+⟨ λ n, f n,
+  λ _ hε, by simpa [norm, norm_def] using f.cauchy hε ⟩
+
+variables (p)
+
+instance complete : cau_seq.is_complete ℤ_[p] norm :=
+⟨ λ f,
+  have hqn : ∥cau_seq.lim (cau_seq_to_rat_cau_seq f)∥ ≤ 1,
+    from padic_norm_e_lim_le zero_lt_one (λ _, norm_le_one _),
+  ⟨ ⟨_, hqn⟩,
+    λ ε, by simpa [norm, norm_def] using cau_seq.equiv_lim (cau_seq_to_rat_cau_seq f) ε⟩⟩
+
+end padic_int
 
 namespace padic_int
 
-variables {p : ℕ} [hp_prime : fact p.prime]
+variables (p : ℕ) [hp_prime : fact p.prime]
 include hp_prime
 
-lemma pow_p_dvd_int_iff (n : ℕ) (a : ℤ) : (p ^ n : ℤ_[p]) ∣ a ↔ ↑p ^ n ∣ a :=
+lemma exists_pow_neg_lt {ε : ℝ} (hε : 0 < ε) :
+  ∃ (k : ℕ), ↑p ^ -((k : ℕ) : ℤ) < ε :=
 begin
-  split,
-  { intro h,
-    cases h with w hw,
-    have := congr_arg has_norm.norm hw,
-    simp only [padic_norm_z.padic_norm_z_of_int, padic_norm_z.norm_p_pow, padic_norm_z.mul,
-               fpow_neg, fpow_coe_nat] at this,
-    rw_mod_cast [padic_norm_e.norm_int_lt_pow_iff_dvd, this],
-    simp only [fpow_neg, fpow_coe_nat, nat.cast_pow],
-    convert mul_le_of_le_one_right _ _ using 1,
-    { apply inv_nonneg.mpr, apply pow_nonneg, exact_mod_cast le_of_lt hp_prime.pos },
-    { apply padic_norm_z.le_one } },
-  { intro h,
-    simpa only [ring_hom.map_pow] using (int.cast_ring_hom ℤ_[p]).map_dvd h, }
+  obtain ⟨k, hk⟩ := exists_nat_gt ε⁻¹,
```

Comment:
There should surely be a lemma saying that if 0<a<1 and b are positive elements of some kind of ordered field (e.g. Q or R -- an archimedean field?), then there's some nat n such that `a^n<b`. Or maybe if a>1 then a^(-n) < b? Theorems like `exists_pow_neg_lt` and `exists_pow_neg_lt_rat` both follow easily from this I guess. My feeling is that these ten lines of proof should be elsewhere.

Reply:
I spent a little bit looking at this. There are lemmas in this direction in `algebra.archimedean`, but there are lots of variants missing. It's a fiddly task to state all the variants with minimal assumptions. If it's alright with the reviewers I'd rather wait and refactor this later -- the to-do list is too long at the moment, and it seems like a waste to do a half-assed job at a refactor right now.